appcelerator/titanium_mobile
# [TIMOB-26586] iOS: Fixed animation bugs with auto-reverse ease-in/ease-out and 180 degree rotation

**JIRA:**
- https://jira.appcelerator.org/browse/TIMOB-26586
- https://jira.appcelerator.org/browse/TIMOB-26590

**Summary:**
- [TIMOB-26586](https://jira.appcelerator.org/browse/TIMOB-26586) Fixed bug where auto-reverse bug with `EASE_IN` and `EASE_OUT`.
  * `EASE_IN` is supposed to `EASE_OUT` when reversed.
  * `EASE_OUT` is supposed to `EASE_IN` when reversed.
  * Not an issue with `EASE_IN_OUT` or `LINEAR` animation curves.
- [TIMOB-26590](https://jira.appcelerator.org/browse/TIMOB-26590) Fixed bug where animation rotation of `-180` would rotate clockwise instead of counter-clockwise.
  * This is a limitation on Apple's end since min/max rotation is -180 and 180 respectively.
  * Only known native work-around for a transformation matrix based animation is to slightly offset from 180.

**Test:**
1. Build and run the below code on iOS.
2. Verify that all squares rotate **clockwise** `180` degrees and then rotate **counter-clockwise** `180` degrees back to their original position.
3. Verify that top-left EASE_IN square starts rotating clockwise slowly and then ends rotating counter-clockwise slowly. _(The bug was it ended quickly.)_
4. Verify that top-right EASE_OUT square starts rotating clockwise quickly and then ends rotating counter-clockwise quickly. _(The bug was it ended slowly.)_
5. Change the below code's `180` value to `-180`.
6. Rebuild and run on iOS.
7. Verify that all squares rotate **counter-clockwise** `180` degrees and then rotate **clockwise** `180` degrees back to their original position.

```javascript
var SQUARE_SIZE = 100;
function createContainer(name, animationCurve) {
	var viewContainer = Ti.UI.createView({
		layout: "vertical",
		width: "50%",
		height: "50%",
	});
	viewContainer.add(Ti.UI.createLabel({
		text: name,
		color: "black",
		top: "5dp",
	}));
	var squareView = Ti.UI.createLabel({
		text: "Test",
		textAlign: Ti.UI.TEXT_ALIGNMENT_CENTER,
		color: "white",
		backgroundColor: "blue",
		top: "15dp",
		width: SQUARE_SIZE,
		height: SQUARE_SIZE,
	});
	var animation = Ti.UI.createAnimation({
		transform: Ti.UI.create2DMatrix().rotate(180),
		curve: animationCurve,
		duration: 1000,
		autoreverse: true,
	});
	viewContainer.add(squareView);
	function doAnimation() {
		squareView.animate(animation, function() {
			setTimeout(function() {
				doAnimation();
			}, 1000);
		});
	}
	doAnimation();
	return viewContainer;
}

var window = Ti.UI.createWindow({
	backgroundColor: "white",
	fullscreen: true,
	layout: "horizontal",
});
window.add(createContainer("EASE_IN", Ti.UI.ANIMATION_CURVE_EASE_IN));
window.add(createContainer("EASE_OUT", Ti.UI.ANIMATION_CURVE_EASE_OUT));
window.add(createContainer("EASE_IN_OUT", Ti.UI.ANIMATION_CURVE_EASE_IN_OUT));
window.add(createContainer("LINEAR", Ti.UI.ANIMATION_CURVE_LINEAR));
window.open();
```
